### ISSUE TITLE
Union types unexpected behavior

### ISSUE THREAD

**Issue by rommeras** [microsoft/TypeScript#41509]
**TypeScript Version:**  4.0.5

**Search Terms:**
array of union types

**Code**

```ts
type A = {a: number; b?: number;}
type B = {c: number; d: number;}
type C = A | B;

const c: Array<C> = [
  {a: 5, c: 6},
];
```

**Expected behavior:**
Type '{ a: number; c: number; }' is not assignable to type 'C'. Property 'd' is missing in type '{ a: number; c: number; }' but required in type 'B'.

**Actual behavior:**
No errors are shown. Making b non optional property of type A gives expected error

**Playground Link:** https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDeBDAXFAdgVwLYCMIAnAbinwH4McDiSBfAKFEigCEFkBjavQ0qABNetUkxbQAwpzgAfdiUaMuAe0wBnYFB6wiRFCAA8kgHycA2oyjJ0UAKwAabRgBs9B4wC6JIA
**Related Issues:**

**Comment by MartinJohns** [comment]
Duplicate of #20863.
